app/models/manager_refresh/graph.rb

```diff
@@ -40,9 +40,22 @@ def to_graphviz(layers: nil)
     def construct_graph!(nodes)
       self.nodes = nodes
       self.edges, self.fixed_edges = build_edges(nodes)
+      assert_edges_in_nodes!
+      assert_fixed_edges_in_edges!
       self
     end
 
+    def assert_edges_in_nodes!
+      edge_endpoints = edges.collect_concat { |e| e }
+      extra = edge_endpoints - nodes
+      raise "Graph has edge endpoints that are not among its nodes: #{extra}" unless extra.empty?
+    end
+
+    def assert_fixed_edges_in_edges!
+      extra = fixed_edges - edges
+      raise "Graph has fixed_edges that are not in edges: #{extra}" unless extra.empty?
+    end
+
     def assert_graph!(fixed_edges)
```

Comment:
@Ladas I'm itching to rename this method, and its argument.
the methods I just added could fit the name "assert graph"; this one is I think `assert_acyclic`?
and the argument — it's called [twice](https://github.com/ManageIQ/manageiq/blob/master/app/models/manager_refresh/inventory_collection/graph.rb#L37-L49), first `assert_graph!(fixed_edges)`, then (after `convert_to_dag!`) with `assert_graph!(edges)`.

and perhaps it's better moved to inventory_collection/graph.rb where all the DAG stuff happens.

Reply:
yeah make sense to rename it to ```assert_acyclic```, the 2 checks are wanted, the first is ```assert_graph!(fixed_edges)``` checking fixed edges, we can't convert that one, that would be e.g. a cycle in unique index columns, so e.g. if the parent_id pointing on the same table would be part of unique index, then we would have to manually save it as a tree walkthrough.

The second assert checks that DAG conversion worked